docs/langref/relay_type.rst

```diff
@@ -0,0 +1,222 @@
+may use the inferred types (e.g., shape information) for making decisions about
+program transformations. For instance, `src/relay/pass/fuse_ops.cc` gives
+an implementation of a pass that uses inferred tensor shapes to replace invocations
+of operators in a Relay program with fused operator implementations.
+
+Reasoning about tensor types in Relay is encoded using type relations, which means
+that the bulk of type checking in Relay is constraint solving (ensuring that all
+type relations are satisfied at call sites). Type relations offer a flexible and
+relatively simple way of making the power of dependent typing available in Relay
+without greatly increasing the complexity of Relay's type system.
+
+Types
+=====
+
+Below we detail the language of types in Relay and how they are assigned to Relay expressions.
+
+Type
+~~~~
+
+The base type for all Relay types. All Relay types are sub-classes of this base type.
+
+See :py:class:`~tvm.relay.ty.Type` for its definition and documentation.
+
+Tensor Type
+~~~~~~~~~~~
+
+A concrete tensor type in Relay.
+
+Tensors are typed according to data type and shape. At present, these use TVM's
+data types and shapes, but in the future, Relay may include a separate AST for
+shapes. In particular, data types include `bool`, `float32`, `int8` and various
+other bit widths. Shapes are given as tuples of dimensions (TVM `IndexExpr`),
+such as `(5, 5)`; scalars are also given tuple types and have a shape of `()`.
+
+Note, though, that TVM shapes can also include variables and arithmetic expressions
+including variables, so Relay's constraint solving phase will attempt to find
+assignments to all shape variables to ensure all shapes will be concrete before
+running a program.
+
+For example, here is a simple concrete tensor type corresponding to a 10-by-10 tensor of 32-bit floats:
+
+.. code-block:: python
+
+   Tensor[(10, 10), float32]
+
+See :py:class:`~tvm.relay.ty.TensorType` for its definition and documentation.
+
+Tuple Type
+~~~~~~~~~~
+
+A type of a tuple in Relay.
+
+Just as a tuple is simply a sequence of values of statically known length, the type
+of the tuple consists of a sequence of the types corresponding to each member of the tuple.
+
+Because a tuple type is of statically known size, the type of a tuple projection
+is simply the corresponding index into the tuple type.
+
+For example, in the below code, :code:`%t` is of type
+`(Tensor[(), bool], Tensor[(10, 10), float32])`
+and :code:`%c` is of type `Tensor[(10, 10), float32]`.
+
+.. code-block:: python
+   let %t = (Constant(0, (), bool), Constant(1, (10, 10), float32));
+   let %c = %t.1;
+   %c
+
+See :py:class:`~tvm.relay.ty.TupleType` for its definition and documentation.
+
+Type Parameter
+~~~~~~~~~~~~~~
+
+Type parameters represent placeholder types used for polymorphism in functions.
+Type parameters are specified according to *kind*: Kinds include general types
```

Comment:
```suggestion
There are four categories, known as *kinds*, of type parameters: :code:`Type`, :code:`BaseType`, :code:`Shape`, and :code:`ShapeVar`. A :code:`Type` type parameter represents the type of a Relay expression (i.e., a subtype of the :code:`Type` node). A :code:`BaseType` is a primitive data type such as :code:`bool`, :code:`float32`. A :code:`Shape` is a tensor shape. A :code:`ShapeVar` is an entry in a tensor shape.
```
There should be examples for each of these. Also the way I've written it conflates the variables with the values they can refer to. That should probably be disambiguated.

Reply:
The trouble is some of these aren't implemented yet, which I will note. I'll see what I could do with regard to examples.